modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,322 @@
+		$lcp_element = $lcp_images_by_minimum_viewport_widths[ $minimum_viewport_widths[ $i ] ];
+		if ( false === $lcp_element || empty( $lcp_element['attributes'] ) ) {
+			// No LCP element at this breakpoint, so nothing to preload.
+			continue;
+		}
+
+		$img_attributes = $lcp_element['attributes'];
+
+		// Prevent preloading src for browsers that don't support imagesrcset on the link element.
+		if ( isset( $img_attributes['src'], $img_attributes['srcset'] ) ) {
+			unset( $img_attributes['src'] );
+		}
+
+		// Add media query.
+		$media_query = sprintf( 'screen and ( min-width: %dpx )', $minimum_viewport_widths[ $i ] );
+		if ( isset( $minimum_viewport_widths[ $i + 1 ] ) ) {
+			$media_query .= sprintf( ' and ( max-width: %dpx )', $minimum_viewport_widths[ $i + 1 ] - 1 );
+		}
+		$img_attributes['media'] = $media_query;
+
+		// Construct preload link.
+		$link_tag = '<link data-ilo-added-tag rel="preload" fetchpriority="high" as="image"';
+		foreach ( array_filter( $img_attributes ) as $name => $value ) {
+			// Map img attribute name to link attribute name.
+			if ( 'srcset' === $name || 'sizes' === $name ) {
+				$name = 'image' . $name;
+			} elseif ( 'src' === $name ) {
+				$name = 'href';
+			}
+
+			$link_tag .= sprintf( ' %s="%s"', $name, esc_attr( $value ) );
+		}
+		$link_tag .= '>';
+
+		$preload_links[] = $link_tag;
+	}
+
+	return implode( "\n", $preload_links );
+}
+
+/**
+ * Walks the provided HTML document, invoking the callback at each open tag.
+ *
+ * @param string   $html              Complete HTML document.
+ * @param callable $open_tag_callback Callback to invoke at each open tag. Callback is passed instance of
+ *                                    WP_HTML_Tag_Processor as well as the breadcrumbs for the current element.
+ * @return string Updated HTML if modified by callback.
+ */
+function ilo_walk_document( string $html, callable $open_tag_callback ): string {
+	$p = new WP_HTML_Tag_Processor( $html );
+
+	/*
+	 * The keys for the following two arrays correspond to each other. Given the following document:
+	 *
+	 * <html>
+	 *   <head>
+	 *   </head>
+	 *   <body>
+	 *     <p>Hello!</p>
+	 *     <img src="lcp.png">
+	 *   </body>
+	 * </html>
+	 *
+	 * The two upon processing the IMG element, the two arrays should be equal to the following:
+	 *
+	 * $open_stack_tags    = array( 'HTML', 'BODY', 'IMG' );
+	 * $open_stack_indices = array( 0, 1, 1 );
+	 */
+	$open_stack_tags    = array();
+	$open_stack_indices = array();
+	while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+		$tag_name = $p->get_tag();
+		if ( ! $p->is_tag_closer() ) {
+
+			// Close an open P tag when a P-closing tag is encountered.
+			if ( in_array( $tag_name, ILO_P_CLOSING_TAGS, true ) ) {
+				$i = array_search( 'P', $open_stack_tags, true );
+				if ( false !== $i ) {
+					array_splice( $open_stack_tags, $i );
+					array_splice( $open_stack_indices, count( $open_stack_tags ) );
+				}
+			}
```

Comment:
There are quite a few more cases of optional closing tags: https://html.spec.whatwg.org/multipage/syntax.html#optional-tags

Nevertheless, given WordPress's legacy of XHTML compatibility, the lack of closing tags may not be common enough to warrant worrying about any of them.